zokrates_core/src/typed_absy/types.rs

```diff
@@ -17,22 +17,27 @@ pub enum Constant<'ast> {
     Concrete(u32),
 }
 
-// At this stage we want all constants to be equal
 impl<'ast> PartialEq for Constant<'ast> {
-    fn eq(&self, _: &Self) -> bool {
-        true
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Constant::Generic(_), _) | (_, Constant::Generic(_)) => true,
+            (Constant::Concrete(a), Constant::Concrete(b)) => a == b,
+        }
     }
 }
 
 impl<'ast> PartialOrd for Constant<'ast> {
-    fn partial_cmp(&self, _: &Self) -> std::option::Option<std::cmp::Ordering> {
-        Some(std::cmp::Ordering::Equal)
+    fn partial_cmp(&self, other: &Self) -> std::option::Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
     }
 }
 
 impl<'ast> Ord for Constant<'ast> {
-    fn cmp(&self, _: &Self) -> std::cmp::Ordering {
-        std::cmp::Ordering::Equal
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match (self, other) {
+            (Constant::Concrete(a), Constant::Concrete(b)) => a.cmp(b),
+            _ => std::cmp::Ordering::Equal,
+        }
```

Comment:
The Hash implementation further down should also be updated or removed for consistency.